feat(video-osd): implement proper OSD design

MrTimscampi opened the pull request:
Based on @Perseusdehond's design.

* The video player now opens maximized by default. This makes it slightly less annoying to use, since most users will watch something instead of continuing to browse.
* Reworks the video OSD to fit the proposed design (With some changes around icons and such)
* Adds Simple Icons to get branded icons across the app (They need to be added to the Nuxt config, due to how Vuetify currently handles custom icons)
* Adds **non-working** "Remote devices", "Playback settings" and "Subtitles" buttons. These are disabled currently and are mostly there to get the groundwork done for further PRs. This PR is already 700 lines, so actually making all that work is best left for multiple followup PRs.
* Adds the queue button to the OSD

![image](https://user-images.githubusercontent.com/19396809/113600138-de342f00-963f-11eb-87b6-7ccfe03b1e11.png)

![image](https://user-images.githubusercontent.com/19396809/113600166-e68c6a00-963f-11eb-95db-a5dcf5f81b7a.png)

![image](https://user-images.githubusercontent.com/19396809/113600206-f0ae6880-963f-11eb-9c6c-478c19bef8dd.png)

![image](https://user-images.githubusercontent.com/19396809/113600225-f906a380-963f-11eb-93fa-a16f3a36be4c.png)

![image](https://user-images.githubusercontent.com/19396809/113600250-002db180-9640-11eb-98db-66b093092820.png)

![image](https://user-images.githubusercontent.com/19396809/113600327-20f60700-9640-11eb-8a81-178751185ad8.png)

neopc10 commented:
Based upon screenshots alone, this looks fantastic. Are the controls off-center?

MrTimscampi commented:
> Based upon screenshots alone, this looks fantastic. Are the controls off-center?

That was indeed the case. I have reworked the way it's handled so that it's properly centered (and added a non-working cast menu, because I get easily distracted).

hamburglar2160 commented:
Looks great, a few things things:
1. The volume slider doesn't appear to be hooked up to the player yet.
2. I think it'd be very handy to have the metadata respond to clicks that will send the player to PIP to the show/season or episode details pages.
3. Season#,Episode# and current/total time text is highlightable.
4. Thoughts on including "ends at:" more or less just above the total time text?

Perseusdehond commented:
It looks really good, have a few things I would change:
1. Is it necessary to have a subtitles button and a settings menu where you can change the subtitles? The subtitles button feels redundant imo.
2. Maybe the pop up menu's shouldn't align to the side of the screen, but to the end of the controls. Now it can feel a bit weird.
3. When a pop up menu is open and you wait too long, the OSD dissapears, but the menu doesn't (this is good). I think the icon linked to that menu should still be visible, otherwise you can't close it easily. / Another way of solving this is closing the menu when you click anywhere else on the screen.
4. Iirc, the previous menu also doesn't close when you open a new one on top.

Guess this was it ;)

MrTimscampi commented:
>     1. The volume slider doesn't appear to be hooked up to the player yet.

It's probably best left for a followup. I really don't want this to balloon, it's already pretty large. It's also somewhat related to a Shaka Player refactoring we've talked about, where we're looking at combining AudioPlayer and VideoPlayer into one generic component, since they're 99% identical.

> I think it'd be very handy to have the metadata respond to clicks that will send the player to PIP to the show/season or episode details pages.

Not planned for now.

> Season#,Episode# and current/total time text is highlightable.

It's known and expected for now. A future PR will disable highlighting app-wide, then re-enable it in specific places.

> Thoughts on including "ends at:" more or less just above the total time text?

I don't really see the usefulness, since you already have the duration of the movie, the current position and the current time. Plus it'd stick out, there's just nowhere to put it.

neopc10 commented on client/locales/en-US.json:
```diff
@@ -197,19 +207,25 @@
   "save": "Save",
   "saved": "Saved",
   "search": "Search",
+  "seasonEpisode": "Season {seasonNumber} Episode {episodeNumber}",
```
I'd add a comma for most locales --- "Season 1, Episode 1"

ThibaultNocchi commented on client/components/Layout/AudioControls.vue:
```diff
@@ -113,7 +113,7 @@
           <v-col cols="3" class="d-none d-md-flex align-center justify-end">
             <like-button :item="getCurrentItem" class="active-button" />
-            <queue-button :item="getCurrentItem" class="active-button" />
+            <queue-button nudge-top="35" />
```
```suggestion
            <queue-button :nudge-top="'35'" />
```
To remove the warning about passing a string to a number prop

ThibaultNocchi commented on client/components/Players/PlayerManager.vue:
```diff
@@ -34,113 +34,108 @@
-                            <v-icon>mdi-closed-caption</v-icon>
+                        <div class="d-flex aligh-center">
+                          <volume-slider class="mr-2" />
+                          <queue-button nudge-top="60" />
+                          <subtitle-selection-button nudge-top="60" />
+                          <playback-settings-button nudge-top="60" />
```
```suggestion
                          <queue-button :nudge-top="'60'" />
                          <subtitle-selection-button :nudge-top="'60'" />
                          <playback-settings-button :nudge-top="'60'" />
```

ThibaultNocchi reviewed (commented):
Two edits to remove warnings/errors in console

neopc10 commented:
Long file name here that'd benefit from a cutoff. Perhaps add some sort of marquee text on mouseover so that the whole name can be made visible?
![image](https://user-images.githubusercontent.com/1365322/114436523-6c09a000-9b93-11eb-96fa-e5d0e7345665.png)

>>I don't really see the usefulness, since you already have the duration of the movie, the current position and the current time. Plus it'd stick out, there's just nowhere to put it.

The ATV and web client both include it and I'd argue that having it on the OSD is as useful as on the item details page.  The positioning here doesn't stick out too much and is similar to what the ATV exoplayer does at the moment. I think bringing down the current/total time to either *just* below the seekbar or as pictured would be worthwhile to make that bar as wide as possible.

>>Is it necessary to have a subtitles button and a settings menu where you can change the subtitles? The subtitles button feels redundant imo.

Agreed here

![image](https://user-images.githubusercontent.com/1365322/114437709-e686ef80-9b94-11eb-81cd-a9b137671cd5.png)

MrTimscampi commented:
>     1. Is it necessary to have a subtitles button and a settings menu where you can change the subtitles? The subtitles button feels redundant imo.

Current version (not pushed yet) hides the subtitles button on very small screens, and hides the subtitle dropdown on larger ones, as a way to save screen space.

> Maybe the pop up menu's shouldn't align to the side of the screen, but to the end of the controls. Now it can feel a bit weird.

It doesn't align with the side of the screen, but tries to open centered over the button. Since there's not enough space, it borders the side of the screen.

> Iirc, the previous menu also doesn't close when you open a new one on top.

Indeed. Though, again, I'd like to avoid ballooning this up too much, since it's already pretty big :yum:

MrTimscampi commented:
Alright, with 93f9754 I consider this good to go, for now.

Added/changed:

* Made the OSD lock from 85ce18d more generic and applied it to all menus on the OSD
* Made the OSD mobile-friendly (The title hides on small screens and elements re-align accordingly)
* The pointer now automatically hides along with the OSD, and shows back up when you move the mouse
* The timeout for the OSD has been adjusted a bit:
  * It's 3 seconds for fine pointers (a mouse)
  * It's 10 seconds for coarse pointers (touchscreens)

This last one is meant to account for the lack of mouse movements on touch screens.

ThibaultNocchi reviewed (approved):
Tried around and reviewed the code

ferferga commented on client/plugins/supportedFeaturesPlugin.ts:
```diff
@@ -31,30 +34,76 @@ declare module 'vuex/types/index' {
-    // Check non-standard Safari PiP support
-    // @ts-expect-error - Non-standard functions doesn't have typings
-    (typeof video.webkitSupportsPresentationMode === 'function' &&
+    const element = document.documentElement;
+
+    return !!(
+      element.requestFullscreen ||
+      // @ts-expect-error -- Non-standard property
+      element.mozRequestFullScreen ||
+      // @ts-expect-error -- Non-standard property
+      element.webkitRequestFullscreen ||
+      // @ts-expect-error -- Non-standard property
+      element.msRequestFullscreen ||
+      // @ts-expect-error -- Non-standard property
+      document.createElement('video').webkitEnterFullscreen
+    );
+  }
```
vue fullscreen already does this for you (check for ``support`` boolean): https://www.npmjs.com/package/vue-fullscreen

ferferga commented on client/components/Buttons/CastButton.vue:
```diff
@@ -0,0 +1,108 @@
+    :z-index="500"
+    class="menu"
+  >
+    <template #activator="{ on: menu, attrs }">
```
No warnings please :(

ferferga commented on client/components/Buttons/PlaybackSettingsButton.vue:
```diff
@@ -0,0 +1,124 @@
+    class="menu"
+    @input="$emit('input', $event)"
+  >
+    <template #activator="{ on: menu, attrs }">
```
No warnings please :(

ferferga commented on client/components/Buttons/SubtitleSelectionButton.vue:
```diff
@@ -0,0 +1,75 @@
+    class="menu"
+    @input="$emit('input', $event)"
+  >
+    <template #activator="{ on: menu, attrs }">
```
No warnings please :(

ferferga commented on client/components/Players/PlayerManager.vue:
```diff
@@ -83,70 +54,134 @@
+                            v-if="$features.fullScreen"
+                            class="align-self-center active-button"
+                            icon
+                            disabled
```
Fullscreen is now possible since #740, please re-add this back

ferferga commented on client/plugins/supportedFeaturesPlugin.ts:
```diff
@@ -31,30 +34,76 @@ declare module 'vuex/types/index' {
+  /**
+   * Detects if the current platform supports showing fullscreen videos
+   *
+   * @returns {boolean}
```
No more warnings please :(

ferferga commented on client/layouts/default.vue:
```diff
@@ -97,6 +97,9 @@
         :fab="!(opaqueAppBar || $vuetify.breakpoint.xsOnly) && !isScrolled"
         bottom
       />
+      <cast-button
```
The buttom needs some margin when not in fab:

![image](https://user-images.githubusercontent.com/10274099/118170059-f7dc4980-b429-11eb-8560-363d3ee68571.png)

![image](https://user-images.githubusercontent.com/10274099/118170148-0e82a080-b42a-11eb-9260-c02e24df326a.png)

MrTimscampi commented on client/store/index.ts:
```diff
@@ -1,6 +1,7 @@
 import Vue from 'vue';
 import { MutationTree, ActionTree } from 'vuex';
-import { UserDto } from '@jellyfin/client-axios';
+import { UserDto } from '@jellyfin/client-axios/models/user-dto';
```
It was an attempt to see if tree-shaking for the API was better this way, but I haven't checked if it did work or not.

MrTimscampi commented on client/nuxt.config.ts:
```diff
@@ -90,7 +91,7 @@ const config: NuxtConfig = {
     'plugins/browserDetectionPlugin.ts',
     { src: 'plugins/playbackProfilePlugin.ts', mode: 'client' },
-    { src: 'plugins/supportedFeaturesPlugin.ts', mode: 'client' },
+    'plugins/supportedFeaturesPlugin.ts',
```
Some of the server-rendered templates and code use $features, so it needs to be loaded on the server-side too.

MrTimscampi commented on client/components/Buttons/CastButton.vue:
```diff
@@ -22,9 +22,8 @@
             :icon="!fab"
             :fab="fab"
-            small
-            disabled
```
@ferferga This menu is non-functional, we shouldn't show it.

ferferga replied:
@MrTimscampi https://matrix.to/#/!kPTHLknOmRaexLxUwT:matrix.org/$YnSGHQIObCdgeWwev5_5YoJ5ObJPIVRHmzUdlOEkfIs?via=matrix.org&via=t2bot.io&via=bonifacelabs.ca

This is the original reason for this, I don't care anyway honestly.
